File: backend/app/schemas/analytics.py
"""
Pydantic schemas for Analytics, Reporting & Dashboards — Phase 9.

All metrics are calculated via PostgreSQL aggregations.
No sensitive data (passwords, tokens, OTPs) is exposed.
"""

from pydantic import BaseModel, ConfigDict


# --------------------------------------------------------------------------- #
# System Overview Schema (Admin only)                                          #
# --------------------------------------------------------------------------- #

class SystemAnalyticsResponse(BaseModel):
    """Global system metrics across users, projects, issues, and severities."""

    total_users: int
    active_users: int
    inactive_users: int
    total_projects: int
    active_projects: int
    total_issues: int
    open_issues: int
    in_progress_issues: int
    resolved_issues: int
    closed_issues: int
    critical_issues: int
    high_issues: int
    medium_issues: int
    low_issues: int

    model_config = ConfigDict(from_attributes=True)


# --------------------------------------------------------------------------- #
# Distribution Schemas                                                         #
# --------------------------------------------------------------------------- #

class IssueStatusDistributionResponse(BaseModel):
    """Issue counts keyed by IssueStatus enum values."""

    REPORTED: int = 0
    TRIAGED: int = 0
    ASSIGNED: int = 0
    IN_DEVELOPMENT: int = 0
    IN_REVIEW: int = 0
    IN_TESTING: int = 0
    QA_VERIFICATION: int = 0
    RESOLVED: int = 0
    CLOSED: int = 0
    REOPENED: int = 0

    model_config = ConfigDict(from_attributes=True)


class SeverityDistributionResponse(BaseModel):
    """Issue counts keyed by Severity enum values."""

    MINOR: int = 0
    MAJOR: int = 0
    CRITICAL: int = 0
    BLOCKER: int = 0

    model_config = ConfigDict(from_attributes=True)


class PriorityDistributionResponse(BaseModel):
    """Issue counts keyed by Priority enum values."""

    LOW: int = 0
    MEDIUM: int = 0
    HIGH: int = 0
    URGENT: int = 0

    model_config = ConfigDict(from_attributes=True)


# --------------------------------------------------------------------------- #
# Trends Schemas                                                               #
# --------------------------------------------------------------------------- #

class IssueTrendItem(BaseModel):
    """Aggregated issue creation and resolution count for a specific date period."""

    date: str
    created_count: int
    resolved_count: int

    model_config = ConfigDict(from_attributes=True)


class IssueTrendResponse(BaseModel):
    """Time-series defect creation and resolution trends."""

    interval: str
    items: list[IssueTrendItem]
    total_created: int
    total_resolved: int

    model_config = ConfigDict(from_attributes=True)


# --------------------------------------------------------------------------- #
# Project Analytics Schemas                                                    #
# --------------------------------------------------------------------------- #

class ProjectAnalyticsResponse(BaseModel):
    """Aggregated issue metrics and resolution rate for a single project."""

    project_id: int
    project_name: str
    project_key: str
    total_issues: int
    open_issues: int
    in_progress_issues: int
    resolved_issues: int
    closed_issues: int
    critical_issues: int
    resolution_rate: float

    model_config = ConfigDict(from_attributes=True)


class ProjectAnalyticsListResponse(BaseModel):
    """List of project analytics items visible to the requester."""

    items: list[ProjectAnalyticsResponse]
    total: int

    model_config = ConfigDict(from_attributes=True)


# --------------------------------------------------------------------------- #
# Developer Performance Schemas (Admin only)                                   #
# --------------------------------------------------------------------------- #

class DeveloperAnalyticsItem(BaseModel):
    """Performance metrics for an individual developer."""

    developer_id: int
    developer_name: str
    developer_email: str
    assigned_issues: int
    resolved_issues: int
    open_issues: int
    resolution_rate: float
    average_resolution_time_hours: float | None = None

    model_config = ConfigDict(from_attributes=True)


class DeveloperAnalyticsResponse(BaseModel):
    """Aggregated performance metrics across all developers in the system."""

    items: list[DeveloperAnalyticsItem]
    total: int

    model_config = ConfigDict(from_attributes=True)


class DeveloperWorkloadItem(BaseModel):
    developer_id: int
    developer_name: str
    developer_email: str
    team: str
    active_tasks: int
    completed_fixes: int
    average_mttr_hours: float | None = None


class DeveloperWorkloadResponse(BaseModel):
    items: list[DeveloperWorkloadItem]
    total: int


# --------------------------------------------------------------------------- #
# Downloadable Report Schema (Admin only)                                      #
# --------------------------------------------------------------------------- #

class AnalyticsReportDataResponse(BaseModel):
    """Aggregated data for generating the downloadable analytics report."""

    system_overview: SystemAnalyticsResponse
    status_distribution: IssueStatusDistributionResponse
    severity_distribution: SeverityDistributionResponse
    priority_distribution: PriorityDistributionResponse
    trends: IssueTrendResponse
    project_analytics: list[ProjectAnalyticsResponse]
    developer_performance: list[DeveloperAnalyticsItem]
    generated_at: str

    model_config = ConfigDict(from_attributes=True)


# --------------------------------------------------------------------------- #
# Quality Metrics Schema (Milestone 2)                                         #
# --------------------------------------------------------------------------- #

class QualityMetricsResponse(BaseModel):
    """
    Computed quality KPIs for the system or a project.

    - fix_rate:             % of issues that have been resolved or closed
    - mttr_hours:           Mean Time To Resolve (hours); None if no resolved issues
    - defect_leakage_rate:  % of critical/blocker issues that were reopened after resolution
    - backlog_health_score: Composite score 0–100 (100 = healthy, 0 = critical backlog)
    - open_critical_count:  Number of CRITICAL or BLOCKER issues still open
    - avg_age_open_days:    Average age of currently open issues in days
    """

    fix_rate_percentage: float
    mean_time_to_resolution_hours: float | None = None
    defect_leakage_rate_percentage: float
    backlog_health_score: float
    open_critical_count: int
    avg_age_open_days: float

    @property
    def fix_rate(self) -> float:
        return self.fix_rate_percentage

    @property
    def mttr_hours(self) -> float | None:
        return self.mean_time_to_resolution_hours

    @property
    def defect_leakage_rate(self) -> float:
        return self.defect_leakage_rate_percentage

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/analytics.ts
export interface SystemAnalyticsResponse {
  total_users: number;
  active_users: number;
  inactive_users: number;
  total_projects: number;
  active_projects: number;
  total_issues: number;
  open_issues: number;
  in_progress_issues: number;
  resolved_issues: number;
  closed_issues: number;
  critical_issues: number;
  high_issues: number;
  medium_issues: number;
  low_issues: number;
}

export interface IssueStatusDistributionResponse {
  REPORTED: number;
  TRIAGED: number;
  ASSIGNED: number;
  IN_DEVELOPMENT: number;
  IN_REVIEW: number;
  IN_TESTING: number;
  QA_VERIFICATION: number;
  RESOLVED: number;
  CLOSED: number;
  REOPENED: number;
}

export interface SeverityDistributionResponse {
  MINOR: number;
  MAJOR: number;
  CRITICAL: number;
  BLOCKER: number;
}

export interface PriorityDistributionResponse {
  LOW: number;
  MEDIUM: number;
  HIGH: number;
  URGENT: number;
}

export interface IssueTrendItem {
  date: string;
  created_count: number;
  resolved_count: number;
}

export interface IssueTrendResponse {
  interval: string;
  items: IssueTrendItem[];
  total_created: number;
  total_resolved: number;
}

export interface ProjectAnalyticsResponse {
  project_id: number;
  project_name: string;
  project_key: string;
  total_issues: number;
  open_issues: number;
  in_progress_issues: number;
  resolved_issues: number;
  closed_issues: number;
  critical_issues: number;
  resolution_rate: number;
}

export interface ProjectAnalyticsListResponse {
  items: ProjectAnalyticsResponse[];
  total: number;
}

export interface DeveloperAnalyticsItem {
  developer_id: number;
  developer_name: string;
  developer_email: string;
  assigned_issues: number;
  resolved_issues: number;
  open_issues: number;
  resolution_rate: number;
  average_resolution_time_hours: number | null;
}

export interface DeveloperAnalyticsResponse {
  items: DeveloperAnalyticsItem[];
  total: number;
}

export interface DeveloperWorkloadItem {
  developer_id: number;
  developer_name: string;
  developer_email: string;
  team: string;
  active_tasks: number;
  completed_fixes: number;
  average_mttr_hours: number | null;
}

export interface DeveloperWorkloadResponse {
  items: DeveloperWorkloadItem[];
  total: number;
}

export interface AnalyticsReportDataResponse {
  system_overview: SystemAnalyticsResponse;
  status_distribution: IssueStatusDistributionResponse;
  severity_distribution: SeverityDistributionResponse;
  priority_distribution: PriorityDistributionResponse;
  trends: IssueTrendResponse;
  project_analytics: ProjectAnalyticsResponse[];
  developer_performance: DeveloperAnalyticsItem[];
  generated_at: string;
}

// -------------------------------------------------------------------------- //
// Quality Metrics — Milestone 2
// -------------------------------------------------------------------------- //

export interface QualityMetricsResponse {
  fix_rate_percentage: number;
  mean_time_to_resolution_hours: number | null;
  defect_leakage_rate_percentage: number;
  backlog_health_score: number;
  open_critical_count: number;
  avg_age_open_days: number;
}

// -------------------------------------------------------------------------- //
// Smart Features — Milestone 3 (Mentor formula)
// -------------------------------------------------------------------------- //

// Priority Calculator — mentor formula: severity_weight × category_urgency_weight
export interface PriorityCalcRequest {
  severity: string;   // CRITICAL | MAJOR | MINOR | TRIVIAL
  category: string;   // Security | Database | API | Backend | UI | Colors | Typo
}

export interface PriorityCalcResponse {
  priority_score: number;
  priority: string;
  severity_weight: number;
  category_urgency_weight: number;
  explanation: string;
  // Legacy fields for backward compat
  recommended_priority: string;
  score: number;
  reasoning: string[];
  confidence: string;
}

// Developer Matcher — mentor spec
export interface DeveloperSuggestion {
  developer_id: number;
  developer_name: string;
  email: string;
  role: string;
  match_percentage: number;
  matched_skills: string[];
  active_task_count: number;
  explanation: string;
  // Legacy fields
  user_id: number;
  full_name: string;
  open_issues: number;
  resolved_issues: number;
  resolution_rate: number;
  average_resolution_time_hours: number | null;
  match_score: number;
  reasons: string[];
}

export interface DeveloperMatchResponse {
  issue_id: number;
  issue_key: string;
  suggestions: DeveloperSuggestion[];
}

// Plotly chart data
export interface PlotlyChartsData {
  defect_trends: {
    dates: string[];
    created: number[];
    resolved: number[];
  };
  severity_distribution: Record<string, number>;
  workflow_pipeline: Record<string, number>;
}
